crates/kitsune_p2p/fetch/src/queue.rs

```diff
@@ -0,0 +1,328 @@
+
+use crate::{FetchContext, FetchKey, FetchOptions, FetchRequest};
+
+/// Max number of queue items to check on each `next()` poll
+const NUM_ITEMS_PER_POLL: usize = 100;
+
+/// A FetchQueue tracks a set of [`FetchKey`]s (op hashes or regions) to be fetched,
+/// each of which can have multiple sources associated with it.
+///
+/// When adding the same key twice, the sources are merged by appending the newest
+/// source to the front of the list of sources, and the contexts are merged by the
+/// method defined in [`FetchQueueConfig`].
+///
+/// The queue items can be accessed only through its Iterator implementation.
+/// Each item contains a FetchKey and one Source agent from which to fetch it.
+/// Each time an item is obtained in this way, it is moved to the end of the list.
+/// It is important to use the iterator lazily, and only take what is needed.
+/// Accessing any item through iteration implies that a fetch was attempted.
+#[derive(Clone)]
+pub struct FetchQueue {
+    config: Arc<dyn FetchQueueConfig>,
+    state: Share<State>,
+}
+
+/// Host-defined details about how the fetch queue should function
+pub trait FetchQueueConfig {
+    /// When a fetch key is added twice, this determines how the two different contexts
+    /// get reconciled.
+    fn merge_contexts(&self, a: u32, b: u32) -> u32;
+}
+
+/// The actual inner state of the FetchQueue, from which items can be obtained
+#[derive(Default, Debug)]
+pub struct State {
+    /// Items ready to be fetched
+    queue: LinkedHashMap<FetchKey, FetchQueueItem>,
+}
+
+/// A mutable iterator over the FetchQueue State
+pub struct StateIter<'a> {
+    state: &'a mut State,
+    interval: Duration,
+}
+
+/// Fetch item within the fetch queue state.
+#[derive(Debug, PartialEq, Eq)]
+struct Sources(Vec<Source>);
+
+#[derive(Debug, PartialEq, Eq)]
+struct FetchQueueItem {
+    /// Known sources from whom we can fetch this item.
+    /// Sources will always be tried in order.
+    sources: Sources,
+    /// The space to retrieve this op from
+    space: KSpace,
+    /// Options specified for this fetch job
+    options: Option<FetchOptions>,
+    /// Opaque user data specified by the host
+    context: Option<FetchContext>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Source {
+    agent: KAgent,
+    last_fetch: Option<Instant>,
+}
+
+impl FetchQueue {
+    /// Add an item to the queue.
+    /// If the FetchKey does not already exist, add it to the end of the queue.
+    /// If the FetchKey exists, add the new source and merge the context in, without
+    /// changing the position in the queue.
+    pub fn push(&mut self, request: FetchRequest, space: KSpace, agent: KAgent) {
+        self.state
+            .share_mut(|s, _| Ok(s.push(&*self.config, request, space, agent)))
+            .expect("no error");
+    }
+
+    /// When an item has been successfully fetched, we can remove it from the queue.
+    pub fn remove(&mut self, key: &FetchKey) {
```

Comment:
I think this needs to return `-> Option<FetchContext>`, so that, when I receive the op-data, if it has been marked as needs validation or whatever we can appropriately pass that info along to conductor.

Reply:
This is implemented.